karabo_data/detector_data.py

```diff
@@ -0,0 +1,393 @@
+                if chunk_tids[-1] < f.train_ids[0] or chunk_tids[0] > f.train_ids[-1]:
+                    # No overlap
+                    continue
+                first_tid = max(chunk_tids[0], f.train_ids[0])
+                first_train_idx = np.nonzero(f.train_ids == first_tid)[0][0]
+                last_tid = min(chunk_tids[-1], f.train_ids[-1])
+                last_train_idx = np.nonzero(f.train_ids == last_tid)[0][0]
+                chunk_firsts = firsts[first_train_idx:last_train_idx+1]
+                chunk_counts = counts[first_train_idx:last_train_idx+1]
+                data_slice = slice(
+                    chunk_firsts[0],
+                    int(chunk_firsts[-1] + chunk_counts[-1]),
+                )
+                trainids = np.repeat(np.arange(first_tid, last_tid+1, dtype=np.uint64),
+                                     chunk_counts.astype(np.intp))
+                pulse_id = f.file['/INSTRUMENT/{}/{}/pulseId'.format(source, group)][data_slice]
+                # Raw files have a spurious extra dimension
+                if pulse_id.ndim >= 2 and pulse_id.shape[1] == 1:
+                    pulse_id = pulse_id[:, 0]
+
+                if isinstance(pulses, by_id):
+                    positions = self._select_pulse_ids(pulses, pulse_id)
+                else:  # by_index
+                    positions = self._select_pulse_indices(
+                        pulses, chunk_firsts - data_slice.start, chunk_counts)
+
+                trainids = trainids[positions]
+                pulse_id = pulse_id[positions]
+                index = pd.MultiIndex.from_arrays([trainids, pulse_id],
+                                                  names=['train', 'pulse'])
+
+                if isinstance(positions, slice):
+                    data_positions = slice(
+                        int(data_slice.start + positions.start),
+                        int(data_slice.start + positions.stop),
+                        positions.step
+                    )
+                else:  # ndarray
+                    # h5py fancy indexing needs a list, not an ndarray
+                    data_positions = list(data_slice.start + positions)
+
+                data = f.file[data_path][data_positions]
+
+                arr = _guess_axes(data, index)
+
+                seq_arrays.append(arr)
+
+        non_empty = [a for a in seq_arrays if (a.size > 0)]
+        if not non_empty:
+            if seq_arrays:
+                # All per-file arrays are empty, so just return the first one.
+                return seq_arrays[0]
+
+            raise Exception(("Unable to get data for source {!r}, key {!r}. "
+                             "Please report an issue so we can investigate")
+                            .format(source, key))
+
+        return xarray.concat(sorted(non_empty,
+                                   key=lambda a: a.coords['train'][0]),
+                            dim='train')
+
+    def get_array(self, key, pulses=by_index[:]):
```

Comment:
It's giving you the pulse IDs; depending on the detector configuration it doesn't necessarily capture every pulse (and IIRC, the machine doesn't necessarily make a real pulse for every pulse ID). When I look at one of the relevant files, `0, 1, 4, 5, 8` does look correct. Try:

```
h5glance CORR-R0034-AGIPD03-S00000.h5 INSTRUMENT/SPB_DET_AGIPD1M-1/DET/3CH0:xtdf/image/pulseId
```

Reply:
oh OK, I get it now.
I have an other question then: I always heard  that this field was not reliably giving the actual (XRAY) pulse id. Did you hear anything else about it?

Reply:
I didn't hear anything about this, but I recall it was something I wanted to check when I was asked to work on code to decode the bunch pattern table. Do you know who you heard it from?

If it's consistently different (e.g. one is twice the other), we can probably work around it for data analysis, but it would be good to have certainty about this, and to know whether & when it will be fixed.

If we can't reliably identify which pulse is which, that seems like a pretty serious problem, and people from CAS/DET/ITDM probably need to be talking to find a solution.

Reply:
Got confirmation (at least for AGIPD) that it is not reliable. Values are given to the detector from a configuration text file...

I don't how easy it would be to work around that, but I fear it is dependent on the detector firmware, which no-one here has control over.

Maybe some other device can figure that out from the machine `bunchPattern` and the detector configuration (acquisition rate, trigger delay, etc.) ...

Reply:
Are the detector settings from that text file available through Karabo, and recorded in the HDF5 files?

If we have all the relevant information, I hope we can match up pulses. But my efforts to get clear documentation for the bunch pattern tables didn't get very far.

Reply:
Path to the configuration file is available in karabo, but not the content, and certainly not stored in files.
I don't have too much knowledge about this, it is something we definitely need to talk to expert  (DET). Maybe we can ask Steffen to join one of the next DA meeting. (and someone from electronic group for the bunch pattern table)

Reply:
sent you an email